### PR TITLE
Fixed #31459 -- Fixed handling invalid indentifiers in URL path conversion.

### PR DESCRIPTION
**Pull request by adamchainz**
[Ticket](https://code.djangoproject.com/ticket/31459).

This found one broken URL in Django's own test suite.

### CODE REVIEW

**Comment by adamchainz** [comment]
This is backwards incompatible, but it seems like quite a minor way to me. URL's with bad path converters syntax (like ticket or the URL in `gis_tests` here) will raise the error on starting Django. Previously they'd be allowed, but I don't think many users will have deliberately used `<` / `>` in URL's. They'll need to convert to `re_path` to continue using them.

I can't think of a way of deprecating this gently apart from parsing URL's with both the old and new regexes. This seems like quite a high price to pay at application startup for such an edge case.

File: django/urls/resolvers.py [hunk 1]
```diff
@@ -273,9 +273,9 @@ def check(self):
         route = self._route
         if '(?P<' in route or route.startswith('^') or route.endswith('$'):
             warnings.append(Warning(
-                "Your URL pattern {} has a route that contains '(?P<', begins "
-                "with a '^', or ends with a '$'. This was likely an oversight "
-                "when migrating to django.urls.path().".format(self.describe()),
+                "Your URL pattern {} has a route that contains begins with a "
+                "'^' or ends with a '$'. This was likely an oversight when "
+                "migrating to django.urls.path().".format(self.describe()),
```
Comment (felixxm) [inline django/urls/resolvers.py:278]: I will revert this changes. Test failure is related with invalid identifier, if we changed `named-group` to the `named_group` all works as expected and the current version of this warning is still valid.

File: docs/releases/3.1.txt [hunk 2]
```diff
@@ -443,6 +443,9 @@ URLs
 * :ref:`Path converters <registering-custom-path-converters>` can now raise
   ``ValueError`` in ``to_url()`` to indicate no match when reversing URLs.
 
+* URL paths using invalid identifiers are now parsed, and will raise an
+  error.
```
Comment (felixxm) [inline docs/releases/3.1.txt:447]: It's an internal cleanup, so release notes are not necessary.

**Comment by felixxm** [comment]
@adamchainz Thanks :+1:

File: django/urls/resolvers.py (hunk 1 above)
Comment (adamchainz) [reply django/urls/resolvers.py]: You're right!

**Comment by adamchainz** [comment]
Thank you for finding a better title for the issue.

**Pull request merged**